### PR TITLE
fix(tz): render all user-visible dates in the restaurant's timezone, not the viewer's

### PR DESCRIPTION
**Pull request by jdelgado2002**
## Why

Users viewing the app from a timezone different from the restaurant's saw dates rendered in **their own** browser timezone. A shift on Monday night in Chicago showed up as Tuesday for a viewer in Auckland; labor cost landed on the wrong business day; schedule conflict detection silently stopped detecting conflicts.

The root cause is a single confusion repeated ~66 times across `src/`: a JavaScript `Date` is either **a day on a calendar** or **a moment in time**, and the two serialize differently. `x.toISOString().split('T')[0]` is neither — it yields the **previous** day for any viewer east of UTC. `new Date('2026-07-15')` is the mirror trap: it parses as UTC midnight and renders as July 14 west of UTC.

## What this does

**1. A clock foundation.** `src/lib/restaurantClock.ts` (moments in time) and `src/lib/dateOnly.ts` (calendar days) — each rejects the other's input rather than silently producing a plausible wrong answer. `useRestaurantClock()` binds them to the selected restaurant's zone.

**2. An ESLint guardrail** (`eslint.config.js`) banning the four patterns that cause this, with a migration allowlist of the 149 files not yet converted. **Shrinking that list is the migration.** Do not add to it.

**3. A timezone cue** shown only when the viewer's zone differs from the restaurant's — no noise for the common case.

**4. The fixes.** The four defect sites from the spec, plus everything the guardrail then exposed.

## Scope grew, deliberately

Turning the guardrail on surfaced **51 pre-existing test failures under `TZ=Pacific/Auckland`** across 15 modules that had never been run outside a US zone. The call was made to fix all of them rather than allowlist them. That roughly doubled the PR.

Two of the nine clusters turned out to be the **tests'** fault, not the source's — those were fixed as test fixtures, and the source was left alone.

Load-bearing bugs found this way that were **not** on the original list:

- **Scheduling let double-bookings through** for any operator east of UTC — `hasScheduleConflict` bucketed shift instants in the viewer's zone, so overlapping shifts compared as non-overlapping.
- **Break-even analysis queried the viewer's date window**, not the restaurant's, via the RPC.
- **Bank re-auth notification emails** rendered dates in the sender's zone; the cohort RPCs had no restaurant timezone to render with (new migration `20260731020000` adds the column).
- **Salary/contractor pay was bucketed by the host clock**, so a period boundary could bill one day at the wrong compensation snapshot.

## Verification

Full suite under three timezones, all identical:

| Zone | Test files | Tests |
|---|---|---|
| `America/Chicago` | 665 passed, 1 skipped | 8187 passed, 2 skipped |
| `Pacific/Auckland` | 665 passed, 1 skipped | 8187 passed, 2 skipped |
| `UTC` | 665 passed, 1 skipped | 8187 passed, 2 skipped |

`npm run typecheck` clean. `npm run build` clean. Numbers are from the tree **after** merging `origin/main` (see below).

## Merged with `main` — two clock modules collided

While this branch was in review, `main` grew its **own** `src/lib/restaurantClock.ts` (#679), plus two more independent timezone fixes (#675, #679). The branch was 7 commits behind and the PR was `CONFLICTING`, which is why no test workflow ever dispatched on it — GitHub can't compute a merge commit for a conflicting PR, so `pull_request` events never fire.

Merged in `20179622`. Seven conflicts. The resolution worth knowing about: `main`'s `restaurantClock.ts` is the more developed sibling (it has an `Intl.DateTimeFormat` cache, `isValidTimezone`, `addDaysToDateStr`, `daysBetweenDateStrs`) but it still had the **UTC-midnight `asInstant` fallback** that this branch fixed in `de0641f8`. Taking `main` wholesale would have silently reverted that fix. The merge takes `main`'s structure and re-applies the zone-aware fallback on top. Full reasoning is in the merge commit message.

## Honest status — please read before approving

- **`npm run lint` does not exit 0.** There are **457** pre-existing eslint errors in `src/`, overwhelmingly `@typescript-eslint/no-explicit-any`. Measured at both ends: **458 at the original merge-base and 458 at HEAD before the merge; 457 after merging `main`, which removed one. This branch adds zero.** `restaurant-clock:` guardrail violations outside the allowlist: **0**.
- **The migration allowlist could not be pruned.** Re-scanned after the merge with the rule forced on: 149 entries, exactly the set of still-violating files. Zero stale, zero unlisted violators.
- **`npm run test:db` (pgTAP) was not run locally** — Docker/local Supabase was down. It runs in CI (`.github/workflows/unit-tests.yml:93`), which is what covers the new migration. The migration was hand-audited: recreated function bodies are byte-identical to the originals apart from the added column and join, grants preserved, and the `NOT EXISTS` subquery alias renamed `r` → `r2` to avoid collision. `restaurants.timezone` confirmed present in production (`text`, nullable, default `'America/Chicago'`).
- **One commit is unrelated to timezones.** `68a709a0` fixes `BulkInventoryDeductionDialog.datePicker.test.tsx`, whose day-cell queries matched adjacent-month filler cells in the calendar grid. It surfaced first under Auckland only because Auckland crosses the month boundary ~17 hours before UTC — it would have turned `main` red for everyone at the month rollover regardless of this branch. Not folded into the timezone narrative.
- **Three test files were inline mirrors** — they reimplemented production logic in the test body instead of importing it, so they were green over a copy and proved nothing. `openShiftCard.test.ts` now imports the real function. The other two mirror Deno edge-function logic that cannot be imported from `src/`; they were verified faithful and in sync.
- **`src/services/laborCalculations.ts`'s `calculateHoursPerEmployee` has no production caller.** The live path is the duplicate at `supabase/functions/_shared/laborCalculations.ts`, which still has the defect. The `src/` fix keeps the suite honest but is not itself a production behavior change. Filed as a follow-up.
- **The Scheduling page header timezone notice** is deliberately deferred to a follow-up PR.

## Review

A whole-branch review on the most capable model returned **0 Critical, 1 Important, 2 Minor**. The Important finding was fixed here in `de0641f8`: `asInstant`'s production fallback built **UTC** midnight, which formats as the previous calendar day in every zone behind UTC — reproducing this branch's own bug class inside its foundation module, on a path that had never been executed in CI. It is now zone-aware, with tests that force the non-throwing production branch and assert `console.error` fired as positive evidence the fallback actually ran.

### CodeRabbit review wave (13 findings, `8a4508e2..d226fd5a`)

Every finding was verified against the code before acting; three of CodeRabbit's proposals turned out to be wrong or incomplete and are called out below.

**Confirmed and fixed** — three of these were defects this branch itself introduced:

| Finding | Fix |
|---|---|
| `openShiftHelpers.hasScheduleConflict` (Critical) | Worse than reported. The same-day pre-filter dropped an employee shift starting the *previous* day, and `toMinutesOfDay` collapses an overnight interval to `start > end`, making the overlap test meaningless. Rewritten to compare **instant intervals**: the open shift's bounds come from `parseWallClock` in the restaurant's zone (rolling to the next day when `end <= start`), employee shifts use their stored instants directly. All 6 existing cases still pass; 3 overnight cases added. |
| `POSSalesImportReview` day shift (Major) | The DatePicker value is a calendar-day token, not an instant — `formatInTimeZone` re-derived a day it never had, shifting imports one day for viewers east of the restaurant. Now `toDateOnlyString`. |
| `usePOSTips` fetch window (Major) | Bucketing had moved to the restaurant's zone but the bounds were bare `YYYY-MM-DD` strings, which Postgres parses against UTC — fetched rows and bucketed rows disagreed. Now `businessDayRangeToInstants` off the same `tz`. |
| `Scheduling` shift-to-column assignment (Major) | The branch had converted the "today" highlight to the restaurant's day while shifts in that same column were still bucketed by the viewer's. Both `selectShiftsForDay` and `getShiftsForEmployee` now use `toBusinessDay`. |
| `TimePunchesManager` displays (Major, partial) | The branch converted the CSV export but left the on-screen punch chips, open-session times, and verification detail on browser-local `format` — export and screen disagreed. Five sites now use `clock.formatInstant`. |
| `payrollCalculations-dailyRate` fixture (Major) | Local-field `new Date(y,m,d)` constructors moved with the host's zone; rebuilt with `Date.UTC(...)`. **No expected value changed** — the splits hold as written under all three matrix zones. |
| Edge-function formatter duplication (Major) | Extracted `supabase/functions/_shared/dateOnly.ts`. There were **three** duplicates, not two — the third was `formatDateLocal` in `_shared/laborCalculations.ts`, which keeps its name (it is imported by name elsewhere, including a dynamic import in `ai-execute-tool`) and now delegates. |
| `InventoryAudit` nested optional chaining (Minor) | Both sites guarded. |
| `useAutomaticInventoryDeduction` (Minor) | Regression test added. |

**Where CodeRabbit was wrong:**

- **`process-receipt` regex.** The escapes *are* unnecessary, but the suggested fix changes `[-_.\/]` to `[-_.]`, which deletes `/` from the character class and stops matching `YYYY/MM/DD` filenames. Applied the correct de-escape instead: `[-_./]`.
- **`compensation-edge-cases` date range — pushed back.** The proposed `2023-12-31..2024-01-06` range would *break* the test: with `hire_date: '2024-01-03'` it yields 4 paid days, but the assertion expects 5, which the existing range produces. The branch only changed `new Date(...)` → `parseDateOnly(...)`; the dates were never wrong. What *was* wrong is the inline label "(Sunday-Saturday)" — 2024-01-01 is a Monday. Corrected the label only.
- **Two findings were raised against a stale revision** (`useRestaurantClock` memo deps, `EmployeeTimecard.periodPunches`) and were already fixed in `c13c4da5` and `402cbc16`.

## Follow-ups filed, not in this PR

- Per-restaurant configurable business-day end (~2 AM)
- App-wide React error boundary (this app has none, which is why `restaurantClock` logs rather than throws in production)
- Fix the edge-function `laborCalculations` duplicate
- The role-resolution / landing-redirect race behind the `permissions-roles` E2E flake
- **`TimePunchesManager`'s fetch window** (`dateRange`, `todaySessions`, header labels) is still viewer-local. Deliberately deferred: it needs the same buffered restaurant-zone treatment payroll got *plus* changes to the `currentDate` navigation state, and unlike the display sites above it is a pre-existing defect rather than one this branch introduced.

### One E2E fix, unrelated to timezones

`tips-flow.spec.ts` went red on shard 4. It is **not** a data or timezone defect: all three failure snapshots show the section rendering `Last 30 days • 1 split`, so the split persisted and displayed. The assertion itself was wrong — `getByText(/recent tip splits/i)` also matches the empty state's two paragraphs, and Playwright fails a multi-element match immediately as a strict mode violation instead of polling. It was a race against the post-approval refetch that happened to have been winning. Now asserts `getByRole('heading', …)`.

One further E2E red, on shard 2, was a **flake and not fixed here**: `permissions-roles.spec.ts:99` (collaborator landing-route redirect). The app code was byte-identical to an earlier run where shard 2 passed — the only diff was the shard-*4* test above — and the three attempts landed on three *different* URLs (`/`, `/recipes`, `/banking`) where one was expected, which is a race, not a fixed routing bug. It passed on rerun. The underlying race between role resolution and the landing redirect is real and filed as a follow-up.

## Known-failing check

`netlify/easyshifthq/deploy-preview` is red, and is **not** caused by this branch: it failed identically on `26d2ef63`, a commit that touched nothing but `.github/workflows/unit-tests.yml`. `npm run build` succeeds locally and Vercel's preview of the same commit deploys clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Dates, times, labor, payroll, scheduling, tips, and timecards now use the restaurant’s configured timezone.
  * Added timezone-aware punch editing, exports, conflict detection, and date-range handling.
  * Added a notice when the viewer’s timezone differs from the restaurant’s timezone.

* **Bug Fixes**
  * Prevented calendar dates from shifting across timezones, including DST transitions and overnight work.

* **Tests**
  * Expanded coverage across UTC, Chicago, and Auckland timezone scenarios.

### CODE REVIEW

File: src/pages/Scheduling.tsx [hunk 1]
```diff
@@ -112,7 +114,8 @@ import {
   Volume2,
   CalendarOff,
 } from 'lucide-react';
-import { format, startOfWeek, endOfWeek, addWeeks, subWeeks, eachDayOfInterval, isSameDay, parseISO, isToday } from 'date-fns';
+import { format, startOfWeek, endOfWeek, addWeeks, subWeeks, eachDayOfInterval, isSameDay, parseISO } from 'date-fns';
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/pages/Scheduling.tsx]: **✅ Agreed** — Correct — startOfWeek was left behind when Scheduling switched to restaurant-zone week bounds. Dropped from the import. Fixed in `b91b2731`.

File: src/hooks/useRestaurantClock.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      tz,
+      tzAbbrev: tzAbbrevOf(tz, now),
+      viewerTzDiffers,
+      today,
+      formatInstant: (value, pattern) => formatInstantIn(value, tz, pattern),
+      toBusinessDay: (value) => toBusinessDayIn(value, tz),
+      toWallClockInput: (value) => toWallClockInputIn(value, tz),
+      parseWallClock: (wallClock) => parseWallClockIn(wallClock, tz),
+    };
+  }, [tz, today]);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/useRestaurantClock.ts]: **✅ Agreed** — Verified real. The memo deps were [tz, today]; today rolls over at the restaurant's midnight, while US DST transitions land at 02:00 local, so from 02:00 until the next restaurant-midnight the memo kept serving the pre-transition abbreviation and offset comparison. tzAbbrev and both offsets are now computed at render time, outside the memo, and threaded into its deps as values — the identity contract still holds (stable across renders when nothing changed) and the staleness window shrinks from 'until tomorrow' to 'until the next render'. Covered by a new test in tests/unit/useRestaurantClock.test.tsx that holds tz and today fixed across the 2026-03-08 spring-forward and asserts CST -> CDT. No timer added: nothing re-renders at the transition itself, which is inherent without polling and deliberately out of scope. Fixed in `c13c4da5`.

File: src/utils/payrollCalculations.ts [hunk 3]
```diff
@@ -489,7 +497,7 @@ export function calculateEmployeePay(
       // OT and tip proration. Same day as startTime for all non-break-crossing
       // shifts, so only the overnight break case changes. clockIn is a required
       // WorkPeriod field (set on every parseWorkPeriods push site).
-      const dateKey = format(new Date(period.clockIn), 'yyyy-MM-dd');
+      const dateKey = toBusinessDay(period.clockIn, timezone);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/utils/payrollCalculations.ts:517]: **✅ Agreed** — Verified real, and it cuts both ways. periodStartDate/periodEndDate are calendar-day tokens (the code a few lines below already serializes them with toDateOnlyString for exactly that reason); the attributeToWindow block was using the same two values as instants, which slides the whole window by the viewer-vs-restaurant offset difference. For a UTC viewer and a Chicago restaurant the window starts too early: the last day's restaurant-evening punches drop out and the evening before the period started is wrongly admitted. Fixed by deriving the window instants from the day tokens in the restaurant's own zone via a new businessDayRangeToInstants helper in src/lib/restaurantClock.ts. Signature unchanged. The same error also corrupted the punch FETCH window, which the review did not mention: bufferPunchFetchRange was buffering the viewer's day bounds in both usePayroll.tsx and EmployeeTimecard.tsx, so at large offsets the boundary-crossing shift was never fetched at all and the pairing engine could not pair it — both now buffer restaurant-zone bounds. Tests: two cases in tests/unit/payrollCalculations.test.ts pinning restaurant-20:00 punches on the period's last day (included) and the day before it starts (excluded) using literal UTC instants, a new bounds assertion in tests/unit/usePayroll.fetchRange.test.ts pinned to literal CST values, and direct coverage of the helper (including the spring-forward day) in tests/unit/restaurantClock.test.ts. The React Query key still uses the logical startDate/endDate, so cache identity is unchanged by the buffer. Fixed in `39ecec38`.

File: src/pages/EmployeeTimecard.tsx [hunk 4]
```diff
@@ -97,24 +96,29 @@ const EmployeeTimecard = () => {
...
+      // serialization here -- do NOT route these through the clock.
+      const dayKey = toDateOnlyString(day);
       grouped.set(dayKey, []);
     });
 
     periodPunches.forEach((punch) => {
-      const punchDate = parseISO(punch.punch_time);
-      const dayKey = format(punchDate, 'yyyy-MM-dd');
+      // punch_time is an instant; bucket it by the restaurant's day.
+      const dayKey = clock.toBusinessDay(punch.punch_time);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/pages/EmployeeTimecard.tsx:143]: **✅ Agreed** — Verified real. periodPunches filtered new Date(punch.punch_time) against the viewer-local startDate/endDate instants, then punchesByDay bucketed the survivors by clock.toBusinessDay, while dayHours was computed from the unfiltered buffered punches — so for any out-of-zone viewer a day could show nonzero hours and 'No punches recorded'. periodPunches now filters on membership in a Set of restaurant business-day keys, the same serialization punchesByDay uses. weekDays is memoized as part of this (it was recomputed every render, which would have defeated the new useMemo), and the fetch window here gets the same restaurant-zone correction as the payroll path. Regression test added in tests/unit/EmployeeTimecard.test.tsx: a 20:00 CDT punch on the period's last day now renders on that card under every zone in the CI matrix. Fixed in `402cbc16`.

File: src/components/POSSalesImportReview.tsx [hunk 5]
```diff
@@ -152,8 +153,9 @@ export const POSSalesImportReview: React.FC<POSSalesImportReviewProps> = ({
     if (!date) return;
     
     setSelectedDate(date);
-    // Get restaurant timezone or fallback to browser timezone
-    const timezone = selectedRestaurant?.restaurant?.timezone || Intl.DateTimeFormat().resolvedOptions().timeZone;
+    // Restaurant timezone, never the viewer's -- an importer in another zone
+    // would otherwise stamp sales with their own calendar day.
+    const timezone = safeTz(selectedRestaurant?.restaurant?.timezone);
     // Format the date in the restaurant's timezone to ensure it stays as the selected date
     const dateString = formatInTimeZone(date, timezone, 'yyyy-MM-dd');
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/components/POSSalesImportReview.tsx]: **✅ Agreed** — Confirmed and branch-introduced. `date` comes from the DatePicker as a local-midnight calendar-day token (case a), so running it through `formatInTimeZone` re-derives a day from an instant it never was — shifting the import by one day for any viewer east of the restaurant. Replaced with `toDateOnlyString(date)`; the now-unused `formatInTimeZone`/`safeTz` imports are gone. New regression test: tests/unit/POSSalesImportReview.applyDate.test.tsx. Fixed in `8a4508e2`.

File: src/hooks/usePOSTips.tsx [hunk 6]
```diff
@@ -35,8 +35,14 @@ function mergeTip(
...
 export function usePOSTips(restaurantId: string | null, startDate: string, endDate: string) {
+  // employee_tips.recorded_at is a moment in time (case b) -- it must bucket
+  // by the RESTAURANT's business day, the same day the pos-tips RPC already
+  // uses, or the two sides of the merge below disagree on which day a tip
+  // belongs to for any viewer/host ahead of the restaurant's timezone.
+  const { tz, toBusinessDay } = useRestaurantClock();
+
   return useQuery({
-    queryKey: ['pos-tips', restaurantId, startDate, endDate],
+    queryKey: ['pos-tips', restaurantId, startDate, endDate, tz],
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/usePOSTips.tsx:46]: **✅ Agreed** — Confirmed. The branch moved bucketing to `toBusinessDay(recorded_at, tz)` but left the fetch bounds as bare 'YYYY-MM-DD' strings, which Postgres parses against UTC — so the rows fetched and the rows bucketed disagreed and edge-of-range tips dropped or leaked. Window now built with `businessDayRangeToInstants(startDate, endDate, tz)` off the same `tz` the bucketing uses, no second source of truth. Test added in tests/unit/posTipsAggregation.test.ts. Fixed in `54f6a937`.

File: src/lib/openShiftHelpers.ts [hunk 7]
```diff
@@ -9,3 +11,45 @@ export function formatCompactTime(time: string): string {
...
+
+  return employeeShifts.some((s) => {
+    if (s.status === 'cancelled') return false;
+    const sDate = toBusinessDay(s.start_time, tz);
+    if (sDate !== openShiftDate) return false;
+    const sStartMin = toMinutesOfDay(formatInstant(s.start_time, tz, 'HH:mm'));
+    const sEndMin = toMinutesOfDay(formatInstant(s.end_time, tz, 'HH:mm'));
+    return sStartMin < osEnd && sEndMin > osStart;
+  });
+}
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/lib/openShiftHelpers.ts]: **✅ Agreed** — Confirmed Critical, and worse than reported — two distinct overnight defects. (1) The `sDate !== openShiftDate` pre-filter dropped an employee shift starting the previous day and running into the open shift's day. (2) `toMinutesOfDay` collapses an overnight interval to start>end, making the overlap test meaningless. Rewrote to compare INSTANT intervals: the open shift's bounds come from `parseWallClock` in the restaurant's zone (rolling to the next day when end<=start), employee shifts use their stored instants directly, overlap is `sStart < oEnd && sEnd > oStart`. All 6 existing cases still pass; 3 overnight cases added in tests/unit/openShiftCard.test.ts. Fixed in `9a757009`.

File: src/pages/Scheduling.tsx [hunk 8]
```diff
@@ -1205,7 +1213,7 @@ const Scheduling = () => {
                           {selectionMode ? (
                             <button
                               type="button"
-                              onClick={() => selectShiftsForDay(format(day, 'yyyy-MM-dd'))}
+                              onClick={() => selectShiftsForDay(toDateOnlyString(day))}
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/pages/Scheduling.tsx:1224]: **✅ Agreed** — Confirmed, and the branch made it internally inconsistent: line 1385's `dayIsToday` was converted to the restaurant's day while the shifts placed in that same column were still bucketed by the viewer's. Fixed both sites — `selectShiftsForDay` (512) and `getShiftsForEmployee` (781) now use `toBusinessDay(s.start_time, restaurantTimezone) === dayKey`; `isSameDay`/`parseISO` dropped from the date-fns import. Test: tests/unit/schedulingDayBucketing.test.ts. Fixed in `8802f119`.

File: src/pages/TimePunchesManager.tsx [hunk 9]
```diff
@@ -74,6 +77,7 @@ type VisualizationMode = 'manual' | 'cards' | 'barcode' | 'stream' | 'receipt';
 const TimePunchesManager = () => {
   const { selectedRestaurant } = useRestaurantContext();
   const restaurantId = selectedRestaurant?.restaurant_id || null;
+  const clock = useRestaurantClock();
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/pages/TimePunchesManager.tsx:80]: **✅ Agreed** — Agreed in part, and split deliberately. The display half is fixed now: the branch had converted the CSV export and edit dialog to the restaurant's zone but left the on-screen punch chips (873-874), open-session clock-in (956), force-clock-out validation message (1089) and verification detail (1142/1146) on browser-local `format(new Date(...))` — so the export and the screen disagreed. All five now use `clock.formatInstant`; the `differenceInMinutes` elapsed-time math and the `.getTime()` comparison are zone-independent and untouched. Test: tests/unit/TimePunchesManager.restaurantZone.test.tsx. The fetch-window half (`dateRange` 285, `todaySessions` 369, header labels 548-552) is NOT in this PR: it needs the same buffered restaurant-zone treatment payroll got plus changes to the `currentDate` navigation state, and it is a pre-existing defect rather than one this branch introduced. Tracked as follow-up work. Fixed in `39069ccd`.

File: src/pages/InventoryAudit.tsx [hunk 10]
```diff
@@ -152,7 +153,7 @@ export default function InventoryAudit() {
 
   const displayValuesMap = useMemo(() => {
     const map = new Map<string, AuditDisplayValues>();
-    const tz = selectedRestaurant?.restaurant.timezone || Intl.DateTimeFormat().resolvedOptions().timeZone;
+    const tz = safeTz(selectedRestaurant?.restaurant.timezone);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/pages/InventoryAudit.tsx]: **✅ Agreed** — Confirmed — the outer `?.` concedes `selectedRestaurant` may be null, then the inner access assumes `restaurant` is always present. Both sites (156 and 219) now use `selectedRestaurant?.restaurant?.timezone`, including the memo dependency. Fixed in `08e92568`.

File: tests/unit/payrollCalculations-dailyRate.test.ts [hunk 11]
```diff
@@ -509,6 +523,7 @@ describe('Payroll Calculations - Daily Rate', () => {
         mockDailyRateEmployee,
         punches,
         0,
+        'UTC',
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/unit/payrollCalculations-dailyRate.test.ts:533]: **✅ Agreed** — Confirmed. The seven-day fixture used local-field `new Date(y,m,d)` constructors, so its instants moved with the host's zone — and this suite now runs under a three-way TZ matrix (America/Chicago, Pacific/Auckland, UTC) in CI. Rebuilt with `Date.UTC(...)`. No expected value changed: the daily-rate splits hold as written under all three zones. Fixed in `a9bc4f70`.

File: supabase/functions/process-expense-invoice/index.ts [hunk 12]
```diff
@@ -143,6 +143,19 @@ function cleanJsonContent(content: string): string {
...
+// viewer/server TZ east of UTC, e.g. Pacific/Auckland). Edge functions are
+// Deno and cannot import `@/lib/dateOnly`, so this is a self-contained
+// version of `toDateOnlyString()` - keep the two in agreement.
+function toDateOnlyLocal(date: Date): string {
+  const year = date.getFullYear();
+  const month = String(date.getMonth() + 1).padStart(2, "0");
+  const day = String(date.getDate()).padStart(2, "0");
+  return `${year}-${month}-${day}`;
+}
+
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/functions/process-expense-invoice/index.ts]: **✅ Agreed** — Confirmed. Extracted supabase/functions/_shared/dateOnly.ts exporting `toDateOnlyString` (local-field serialization, mirroring src/lib/dateOnly.ts) and pointed the duplicates at it. There were three, not two — process-expense-invoice, process-receipt, and `formatDateLocal` in _shared/laborCalculations.ts. `formatDateLocal` keeps its name because it is imported by name elsewhere, including a dynamic import in ai-execute-tool; it now delegates. Verified with `deno check` (edge functions are outside tsconfig.app.json's include). Test: tests/unit/edgeDateOnly.test.ts. Fixed in `4d70e8c3`.

File: src/hooks/useAutomaticInventoryDeduction.tsx [hunk 13]
```diff
@@ -102,9 +105,12 @@ export const useAutomaticInventoryDeduction = () => {
...
     try {
-      // Get sales from today that haven't been processed
-      const today = new Date().toISOString().split('T')[0];
-      
+      // Get sales from today that haven't been processed. sale_date is a
+      // date-only column keyed to the restaurant's business day; the UTC day
+      // (toISOString) rolls over 5-6 hours early for US zones, which silently
+      // stops auto-deduction during evening service.
+      const today = toBusinessDay(new Date(), tz);
+
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/useAutomaticInventoryDeduction.tsx:113]: **✅ Agreed** — Agreed — added tests/unit/useAutomaticInventoryDeduction.test.ts pinning that a sale instant is attributed to the restaurant's business day when the viewer's local day differs. Fixed in `d226fd5a`.

File: supabase/functions/process-receipt/index.ts [hunk 14]
```diff
@@ -52,35 +52,48 @@ function parsePurchaseDate(dateString: string | undefined): string | null {
...
 function extractDateFromFilename(filename: string | null): string | null {
   if (!filename) return null;
-  
+
   // Remove file extension
   const nameWithoutExt = filename.replace(/\.[^/.]+$/, '');
-  
+
   // Pattern 1: YYYY-MM-DD or YYYY_MM_DD or YYYY.MM.DD
   const isoPattern = /(\d{4})[-_.\/](\d{1,2})[-_.\/](\d{1,2})/;
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/functions/process-receipt/index.ts]: **✅ Agreed** — Agreed that the escapes are unnecessary — but the suggested fix is wrong and was not applied. The prompt proposes changing `[-_.\/]` to `[-_.]`, which deletes `/` from the character class entirely and stops matching YYYY/MM/DD filenames. The correct de-escape keeps the slash: both patterns are now `[-_./]`. Fixed in `08e92568`.

File: tests/unit/compensation-edge-cases.test.ts [hunk 15]
```diff
@@ -648,11 +649,14 @@ describe('Salaried Employee Edge Cases', () => {
...
+      // Calculate pay for full week (Sunday-Saturday).
+      // parseDateOnly anchors at LOCAL midnight (unlike bare `new Date(str)`,
+      // which parses date-only strings as UTC midnight and would shift the
+      // loop's local calendar day for viewers behind UTC).
       const fullWeekPay = calculateSalaryForPeriod(
         employee,
-        new Date('2024-01-01'), // Sunday (before hire)
-        new Date('2024-01-07')  // Saturday
+        parseDateOnly('2024-01-01'), // Sunday (before hire)
+        parseDateOnly('2024-01-07')  // Saturday
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply tests/unit/compensation-edge-cases.test.ts]: **↩️ Pushed back** — Pushing back on the date change; taking the comment fix. The proposed 2023-12-31..2024-01-06 range would break the test: with `hire_date: '2024-01-03'` it yields 4 paid days, but the assertion expects 5 ('Should only pay for 5 days'), which is exactly what the existing 2024-01-01..2024-01-07 range produces. The branch only changed `new Date(...)` to `parseDateOnly(...)` here — the dates were never wrong. What IS wrong is the inline label '(Sunday-Saturday)': 2024-01-01 is a Monday. Corrected the label to '(Monday-Sunday)' and left the dates and assertions as they were.

File: src/hooks/useRestaurantClock.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/hooks/useRestaurantClock.ts]: **✅ Agreed** — Already fixed earlier on this branch, in c13c4da5 — the finding was raised against a stale revision. `abbrev`, `viewerOffsetMinutes` and `restaurantOffsetMinutes` are all in the memo's dependency array now (`[tz, today, abbrev, viewerOffsetMinutes, restaurantOffsetMinutes]`), with a comment explaining why `today` alone is insufficient: across a DST transition `tz` and `today` are both unchanged, so a memo keyed only on those would serve a pre-transition `tzAbbrev`/`viewerTzDiffers` for up to ~22h. Fixed in `c13c4da5`.

File: src/pages/EmployeeTimecard.tsx [hunk 16]
```diff
@@ -97,24 +96,29 @@ const EmployeeTimecard = () => {
...
       if (grouped.has(dayKey)) {
         grouped.get(dayKey)!.push(punch);
       }
     });
 
     return grouped;
-  }, [periodPunches, weekDays]);
+  }, [periodPunches, weekDays, clock]);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/pages/EmployeeTimecard.tsx:150]: **✅ Agreed** — Already fixed earlier on this branch, in 402cbc16 — raised against a stale revision. `periodPunches` filters by restaurant business-day membership (`weekDayKeys.has(clock.toBusinessDay(punch.punch_time))`), matching how punchesByDay/dayHours bucket. Covered by tests/unit/EmployeeTimecard.test.tsx, which pins the case of a punch late in the restaurant's business day falling after the viewer-local endDate instant. Fixed in `402cbc16`.

File: src/lib/openShiftHelpers.ts [hunk 17]
```diff
@@ -9,3 +11,49 @@ export function formatCompactTime(time: string): string {
...
+    openEndHHMM > openStartHHMM ? openShiftDate : addDaysToDateStr(openShiftDate, 1);
+
+  const oStart = new Date(parseWallClock(`${openShiftDate}T${openStartHHMM}`, tz)).getTime();
+  const oEnd = new Date(parseWallClock(`${openEndDate}T${openEndHHMM}`, tz)).getTime();
+
+  return employeeShifts.some((s) => {
+    if (s.status === 'cancelled') return false;
+    const sStart = new Date(s.start_time).getTime();
+    const sEnd = new Date(s.end_time).getTime();
+    return sStart < oEnd && sEnd > oStart;
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply src/lib/openShiftHelpers.ts:57]: **↩️ Pushed back** — The schema forbids the state this guards against: shifts.start_time and shifts.end_time are TIMESTAMPTZ NOT NULL with CHECK (end_time > start_time) (supabase/migrations/20251114100000_create_scheduling_tables.sql:22-30), and the only caller (src/pages/AvailableShiftsPage.tsx:266) feeds hasScheduleConflict rows straight from useShifts on that table. A Number.isNaN guard here would be unreachable branch we could not write a meaningful test for, so we are leaving the types as the contract.

File: supabase/functions/_shared/laborCalculations.ts [hunk 18]
```diff
@@ -145,11 +147,12 @@ const MAX_SHIFT_GAP_HOURS = 18;
...
+// Kept as `formatDateLocal` (rather than renamed to `toDateOnlyString`) because
+// it's imported by name elsewhere, including a dynamic import in
+// ai-execute-tool/index.ts. Delegates to the shared Deno-side implementation
+// so this module and process-receipt/process-expense-invoice don't drift.
 export function formatDateLocal(date: Date): string {
-  const year = date.getFullYear();
-  const month = String(date.getMonth() + 1).padStart(2, '0');
-  const day = String(date.getDate()).padStart(2, '0');
-  return `${year}-${month}-${day}`;
+  return toDateOnlyString(date);
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply supabase/functions/_shared/laborCalculations.ts:155]: **↩️ Pushed back** — Real bug, but pre-existing and out of this PR's scope. This branch's only change to formatDateLocal was a pure de-duplication: it replaced an inline getFullYear/getMonth/getDate formatter with the identical toDateOnlyString from _shared/dateOnly.ts so this module and process-receipt/process-expense-invoice stop drifting. Bucketing punch_time by the restaurant's zone instead of the edge runtime's is a behaviour change across every caller of parseWorkPeriods, and this PR is scoped to the client-side restaurantClock foundation, the ESLint guardrail, and four named defect sites. It is tracked as its own follow-up.

File: .github/workflows/unit-tests.yml [hunk 19]
```diff
@@ -50,6 +46,47 @@ jobs:
...
+          cache: 'npm'
+
+      - name: Install dependencies
+        run: npm ci
+
+      - name: Run unit tests under ${{ matrix.tz }}
+        run: npm run test
+        env:
+          TZ: ${{ matrix.tz }}
+
```
[the remark this answers is not in the document]
Comment (jdelgado2002) [reply .github/workflows/unit-tests.yml:93]: **✅ Agreed** — Correct — the timezone-matrix job does no git write and npm ci runs third-party install scripts, so there is no reason to leave GITHUB_TOKEN in the git config for them. Added persist-credentials: false to that checkout step. Fixed in `d74a15c3`.
